emma-language/src/main/scala/org/emmalanguage/compiler/lang/comprehension/Combination.scala

```diff
@@ -442,29 +451,33 @@ private[comprehension] trait Combination extends Common {
       } yield {
         // It can happen that the key functions have differing return types (e.g. Long and Int).
         // In this case, we add casts before the return expr of the key functions to the weakLub.
-        def maybeCast(tree: u.Tree) = if (kxExpr.tpe =:= kyExpr.tpe) tree else tree match {
+        val kLub = api.Type.weakLub(Seq(kxExpr.tpe, kyExpr.tpe))
+        def maybeCast(tree: u.Tree) = if (tree.tpe <:< kLub) tree else tree match {
```

Comment:
Subtypes are ok, because functions are covariant in the return type. However calling `asInstanceOf` is still necessary in case of promoting primitive types (e.g. `Int` to `Double`), because they are not in a subtype relationship.